LunrCore/QueryParser.cs

```diff
@@ -45,11 +45,11 @@ public Query Parse()
             return Query;
         }
 
-        private Lexeme PeekLexeme()
-            => _lexemeIndex < _lexemes.Count ? _lexemes[_lexemeIndex] : Lexeme.Empty;
+        private Lexeme? PeekLexeme()
+            => _lexemeIndex < _lexemes.Count ? _lexemes[_lexemeIndex] : null;
```

Comment:
Not sure how I feel about that. I've been deliberately trying to avoid using null, and use more expressive static empty instances. What's wrong with that?

Reply:
This feels like the proper use of null as we're signaling a EOF that must be handled. 

This also forces consumers to ensure a non-null state before checking the members. We get this for free by enabling nullable and treating warnings as errors.

The behavior of the "==" operator also isn't clear at the call site -- since these can be overwritten by the class. Comparing using a is null pattern is always guaranteed to have the same behavior.

Reply:
Well, null is the complete absence of semantic contents, by definition. If we're signaling an EOF condition, let's reflect that explicitly and expressively, but not with null.

Reply:
Indeed. If only we had discriminated unions and could do `Lexeme | EndOfString`. In absence of an obvious pattern here (Try doesn't feel right), will back out on this change too.

I'm current experimenting with a few bigger changes that have a measurable impact. Will hold off any further changes without presenting evidence of the improvement.